---
title: "Mcap Developmental pH Series of Symbiont Gene Expression in *M. capitata*"
author: "Erin Chille"
date: "Last updated 2020/11/12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE) #Set Strings to character
```

# BSF Biomineralization Project: RNAseq Expression and Functional Analysis

## Set up workspace

Import necessary libraries
```{r, message=FALSE, warning=FALSE}
library("tidyverse")
library("tximport")
library("genefilter")
library("DESeq2")
library("RColorBrewer")
library("gridExtra")
library("ComplexHeatmap")
library("dplyr")
```

## Data input, cleaning, and pre-processing

Import the data files 
```{r}
treatmentinfo <- read.csv("Sample_Info/RNAseq_data.csv", header = TRUE, sep = ",", row.names = "sample_id")
head(treatmentinfo)

gcount <- as.data.frame(read_delim(file = "6-Symbiont-RNAseq-Analysis/a-QC-Align-Assemble/Output/gene_counts_matrix.tsv", delim = "\t"))
rownames(gcount) <- gcount$X1
gcount$X1 <- NULL
head(gcount)
nrow(gcount)
gcount <- round(gcount) # round the gene counts matrix

treatmentinfo <- treatmentinfo[colnames(gcount),] #remove samples not in the gcount matrix from treatment info csv 
```
## Pre-filter gene counts
Pre-filtering our dataset to reduce the memory size dataframe, increase the speed of the transformation and testing functions, and improve quality of statistical analysis by removing low-coverage counts. Removed counts could represent outliers in the data and removing these improves sensitivity of statistical tests. Here we will filter out the samples from time points where we only took ambient samples.. We will also apply a filter so only transcripts with low coverage using the PoverA filter described above (P=0.875, A=10). 
```{r}
filt <- filterfun(pOverA(0.875,10)) #filter out genes that have a count of 5 or fewer in less than 1/9 samples

#create filter for the counts data
gfiltpln <- genefilter(gcount, filt)
#identify genes to keep by count filter
gkeeppln <- gcount[gfiltpln,]
#identify gene lists
gn.keeppln <- rownames(gkeeppln)
#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt_pln <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keeppln),])
nrow(gcount_filt_pln)
```
#### Construct the DESeq dataset
Merge the treatment and time_point columns into a new column , group. Set group as a factor.
```{r}
treatmentinfo$treatment <- factor(treatmentinfo$treatment, levels = c("amb","l", "xl"))
head(treatmentinfo)
head(gcount_filt_pln)
```

Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at the interaction of time_point and treatment to test for any differences in gene expression across timepoints attributed to treatment.
```{r}
#Set DESeq2 design
gdds_pln <- DESeqDataSetFromMatrix(countData = gcount_filt_pln,
                              colData = treatmentinfo,
                              design = ~treatment)
```

#### Visualize gene count data

We're looking to see if the samples of the same pH treatments cluster

##### Log-transform the count data

Log-transform the data using a variance stabilizing transforamtion (vst) for visualization purposes, only. To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4. Otherwise, there could be artefacts in our results.
```{r}
SF.gdds_pln <- estimateSizeFactors( gdds_pln ) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than for to use vst
print(sizeFactors(SF.gdds_pln)) #View size factors
```

Our size factors are all less than 4, so we can use VST!
```{r}
gvst_pln <- varianceStabilizingTransformation(gdds_pln, blind=TRUE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
```

#### Principal component plot of samples
```{r}
gPCAdata_pln <- plotPCA(gvst_pln, intgroup = c("treatment"), returnData=TRUE)
percentVar_pln <- round(100*attr(gPCAdata_pln, "percentVar")) #plot PCA of samples with all data
ggplot(gPCAdata_pln, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar_pln[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pln[2],"% variance")) +
  scale_color_manual(values = c(amb="slateblue", l="indianred3", xl="deeppink4")) +
  coord_fixed() +
    theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) + #Set the plot background
  theme(legend.position = ("none")) #set title attributes
```

There appears to be clustering by treatment, especially on the PC1 axis.

#### Differential Gene Expression Analysis

##### Run DE analysis

Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEG_pln <- DESeq(gdds_pln) #run differential expression test by group using the Wald model
```

Explore significant p-values for treatment_l_vs_amb, treatment_xl_vs_amb, and treatment_xl_vs_l
```{r, message = FALSE}
DEG_pln.results.l_vs_amb <- results(DEG_pln, contrast=c("treatment","l","amb"))
summary(DEG_pln.results.l_vs_amb) #view results summary
sum(DEG_pln.results.l_vs_amb$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

DEG_pln.results.xl_vs_amb <- results(DEG_pln, contrast=c("treatment","xl","amb"))
summary(DEG_pln.results.xl_vs_amb) #view results summary
sum(DEG_pln.results.xl_vs_amb$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

DEG_pln.results.xl_vs_l <- results(DEG_pln, contrast=c("treatment","xl","l"))
summary(DEG_pln.results.xl_vs_l) #view results summary
sum(DEG_pln.results.xl_vs_l$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

#How many adjusted p-values were less than 0.05 and what were they?
LvAMB <- filter(as.data.frame(DEG_pln.results.l_vs_amb), padj<0.05)
XLvAMB <- filter(as.data.frame(DEG_pln.results.xl_vs_amb), padj<0.05)
XLvL <- filter(as.data.frame(DEG_pln.results.xl_vs_l), padj<0.05)

LvAMB$contrast <- "LvAMB"
LvAMB$gene_id <- rownames(LvAMB)
rownames(LvAMB) <- seq(1:12)

XLvAMB$contrast <- "XLvAMB"
XLvAMB$gene_id <- rownames(XLvAMB)
rownames(XLvAMB) <- seq(from = 13, to = 29, by = 1)
blast_DEGs <- bind_rows(LvAMB, XLvAMB)
```
There are a fews DEGs in the symbionts at the planula stage.

Get fasta for blast
```{r}
# library("phylotools")
# fasta <- read.fasta(file = "6-Symbiont-RNAseq-Analysis/a-QC-Align-Assemble/Output/symb_Trinity.fasta", clean_name = FALSE) #load fasta file
# fasta$seq.name <- as.factor(fasta$seq.name)
# rownames(fasta) <- fasta$seq.name
# nXA <- fasta[rownames(XLvAMB),]
# nLA <- fasta[rownames(LvAMB),]
# nXLL <- fasta[which(fasta$seq.name%in%rownames(XLvL))]
# dat2fasta(na.omit(nXA), "6-Symbiont-RNAseq-Analysis/b-Differential-Gene-Expression-Analysis/Output/degsXLvAMB.fasta")
# dat2fasta(na.omit(nLA), "6-Symbiont-RNAseq-Analysis/b-Differential-Gene-Expression-Analysis/Output/degsLvAMB.fasta")
# #dat2fasta(na.omit(fasta[intersect(nLA$seq.name, nXA$seq.name),]), "6-Symbiont-RNAseq-Analysis/b-Differential-Gene-Expression-Analysis/Output/SHAREDdegsLXLvAMB.fasta")

```
#Make a heatmap of DEGs
Subset the gene count matrix by the list of DEGs
```{r}
LvAMB$gene_id <- rownames(LvAMB)
XLvAMB$gene_id <- rownames(XLvAMB)
XLvL$gene_id <- rownames(XLvL)
pln_DEGs_all <- bind_rows(LvAMB, XLvAMB, XLvL)
pln_DEGlist <- gdds_pln[unique(pln_DEGs_all$gene_id),]
```

We know from before that our size factors are all less than 4, so we can use VST!
```{r}
pln_DEGvst <- vst(pln_DEGlist, blind=FALSE, nsub = nrow(counts(pln_DEGlist))) #apply a variance stabilizing transforamtion to minimize efplncts of small counts and normalize wrt library size
head(assay(pln_DEGvst)) #view transformed gene count data
dim(assay(pln_DEGvst))
```

Make a matrix for computing similarity
```{r}
matpln <- assay(pln_DEGvst)#[pln_topVarGenes, ] #make an expression object
matpln <- matpln - rowMeans(matpln) #difference in expression compared to average across all samples
```
Write plot
```{r}
#Prepare annotations
hm_ann_col <- HeatmapAnnotation(df=as.data.frame(colData(pln_DEGvst)[c("treatment")]), col = list(treatment=c(amb="slateblue", l="indianred3", xl="deeppink4"))) #make dataframe for column naming
pln_DEGheatmap <-  Heatmap(matpln, column_title = "Treatment", 
        row_title = "DEGs", name = "expression",
        show_row_names = TRUE, top_annotation = hm_ann_col, show_column_names = TRUE, row_dend_side = "left" ,
        column_split = 2, column_dend_height = unit(0.5, "in"),
        #km = 2, row_km_repeats = 100, 
        row_gap = unit(2.5, "mm"), border = TRUE,
        column_names_gp =  gpar(fontsize = 10)); pln_DEGheatmap
# png(file = "4-Differential-Gene-Expression-Analysis/Output/OA-PLN_DEGheatmap.png")
# pln_DEGheatmap
# dev.off()
```

Which are down-regulated in extreme low vs low and ambient
```{r}
down <- c("TRINITY_DN1741_c0_g1","TRINITY_DN13136_c1_g1","TRINITY_DN140660_c0_g1","TRINITY_DN1741_c1_g1","TRINITY_DN220095_c0_g1")
filter(XLvAMB, gene_id%in%down)
filter(XLvL, gene_id%in%down)
filter(LvAMB, gene_id%in%down)
```

```{r}
up <- c("TRINITY_DN2646_c0_g1", "TRINITY_DN20158_c0_g2","TRINITY_DN454_c1_g1","TRINITY_DN5759_c3_g1","TRINITY_DN45797_c0_g2","TRINITY_DN1625_c0_g1","TRINITY_DN120786_c0_g1")
filter(XLvAMB, gene_id%in%up)
filter(XLvL, gene_id%in%up)
filter(LvAMB, gene_id%in%up)
```
# Run BLAST to determine if photosynthesis-related genes are in our *de novo* reference and are being included in our DESeq2 run

Run blast
```{bash}
#Make blast database from Mcap predicted protein fasta file
#/usr/local/ncbi/blast/bin/makeblastdb -in 6-Symbiont-RNAseq-Analysis/a-QC-Align-Assemble/Output/symb_Trinity.fasta -input_type fasta -dbtype nucl -parse_seqids -out 6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/symb.nucl.blast.db

#Blast fasta file containing sequences of selected biomarkers against our sequences
#/usr/local/ncbi/blast/bin/tblastn -query 6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/photosynthesis.prot.fasta -db 6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/symb.nucl.blast.db -outfmt 6 -max_target_seqs 100 -evalue 1e-05 -out 6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/symb_photo_biomarker_hits.prot.csv

#/usr/local/ncbi/blast/bin/blastn -query 6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/photosynthesis.nucl.fasta -db 6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/symb.nucl.blast.db -outfmt 6 -max_target_seqs 100 -evalue 1e-05 -out 6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/symb_photo_biomarker_hits.nucl.csv
```

Examine results

```{r}
hits.nucl <- read.table(file="6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/symb_photo_biomarker_hits.nucl.csv", col.names=c("Enzyme_Name","gene_id","pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "ssend", "evalue", "bitscore"))
hits.nucl <- hits.nucl %>% distinct(Enzyme_Name, gene_id, .keep_all = TRUE)
hits.nucl <- hits.nucl %>% group_by(Enzyme_Name) %>% distinct(sstart, ssend, .keep_all = TRUE)
hits.nucl$gene_id <- gsub("augustus.", "", hits.nucl$gene_id)
hits.nucl$gene_id <- gsub(".t1", "", hits.nucl$gene_id)
levels(as.factor(hits.nucl$Enzyme_Name)) #there should be 5

hits.prot <- read.table(file="6-Symbiont-RNAseq-Analysis/c-Blast-DEGs/symb_photo_biomarker_hits.prot.csv", col.names=c("Enzyme_Name","gene_id","pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "ssend", "evalue", "bitscore"))
hits.prot <- hits.prot %>% distinct(Enzyme_Name, gene_id, .keep_all = TRUE)
hits.prot <- hits.prot %>% group_by(Enzyme_Name) %>% distinct(sstart, ssend, .keep_all = TRUE)
hits.prot$gene_id <- gsub("augustus.", "", hits.prot$gene_id)
hits.prot$gene_id <- gsub(".t1", "", hits.prot$gene_id)
levels(as.factor(hits.prot$Enzyme_Name)) #there should be 2

photo_genes <- bind_rows(hits.nucl, hits.prot) #combine into 1 df
photo_genes$gene_id <- gsub("_i\\d", "", photo_genes$gene_id) #remove isoforms because they are not in our counts matrix and results tables
photo_genes <- photo_genes[!duplicated(photo_genes$gene_id), ] #collapse so there is 1 hit per gene_id
```

See if any of the hits are in any of the results tables and the filtered counts matrix.
```{r}
a <- as.data.frame(DEG_pln.results.l_vs_amb)
a$gene_id <- rownames(b)
b <- as.data.frame(DEG_pln.results.xl_vs_amb)
b$gene_id <- rownames(b)
c <- as.data.frame(DEG_pln.results.xl_vs_l)
c$gene_id <- rownames(c)

filter(a, gene_id%in%photo_genes$gene_id)
filter(b, gene_id%in%photo_genes$gene_id)
filter(c, gene_id%in%photo_genes$gene_id)

gcount_filt_pln$gene_id <- rownames(gcount_filt_pln)
filter(gcount_filt_pln, gene_id%in%photo_genes$gene_id)
```

It appears that these at least some of genes are included!