---
title: "Mcap Developmental pH Series of Symbiont Gene Expression in *M. capitata*"
author: "Erin Chille"
date: "Last updated 2020/11/12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# The following setting is important, do not omit.
options(stringsAsFactors = FALSE) #Set Strings to character
```

# BSF Biomineralization Project: RNAseq Expression and Functional Analysis

## Set up workspace

Import necessary libraries
```{r, message=FALSE, warning=FALSE}
library("tidyverse")
library("tximport")
library("genefilter")
library("DESeq2")
library("RColorBrewer")
library("gridExtra")
library("dplyr")
library("phylotools")
```

## Data input, cleaning, and pre-processing

Import the data files 
```{r}
treatmentinfo <- read.csv("Sample_Info/RNAseq_data.csv", header = TRUE, sep = ",", row.names = "sample_id")
head(treatmentinfo)

gcount <- as.data.frame(read_delim(file = "6-Symbiont-RNAseq-Analysis/a-QC-Align-Assemble/Output/gene_counts_matrix.tsv", delim = "\t"))
rownames(gcount) <- gcount$X1
gcount$X1 <- NULL
head(gcount)
nrow(gcount)
gcount <- round(gcount) # round the gene counts matrix

treatmentinfo <- treatmentinfo[colnames(gcount),] #remove samples not in the gcount matrix from treatment info csv 
```
## Pre-filter gene counts
Pre-filtering our dataset to reduce the memory size dataframe, increase the speed of the transformation and testing functions, and improve quality of statistical analysis by removing low-coverage counts. Removed counts could represent outliers in the data and removing these improves sensitivity of statistical tests. Here we will filter out the samples from time points where we only took ambient samples.. We will also apply a filter so only transcripts with low coverage using the PoverA filter described above (P=0.875, A=10). 
```{r}
filt <- filterfun(pOverA(0.875,10)) #filter out genes that have a count of 5 or fewer in less than 1/9 samples

#create filter for the counts data
gfiltpln <- genefilter(gcount, filt)
#identify genes to keep by count filter
gkeeppln <- gcount[gfiltpln,]
#identify gene lists
gn.keeppln <- rownames(gkeeppln)
#gene count data filtered in PoverA, P percent of the samples have counts over A
gcount_filt_pln <- as.data.frame(gcount[which(rownames(gcount) %in% gn.keeppln),])
nrow(gcount_filt_pln)
```
#### Construct the DESeq dataset
Merge the treatment and time_point columns into a new column , group. Set group as a factor.
```{r}
treatmentinfo$treatment <- factor(treatmentinfo$treatment, levels = c("amb","l", "xl"))
head(treatmentinfo)
head(gcount_filt_pln)
```

Create a DESeqDataSet design from gene count matrix and labels. Here we set the design to look at the interaction of time_point and treatment to test for any differences in gene expression across timepoints attributed to treatment.
```{r}
#Set DESeq2 design
gdds_pln <- DESeqDataSetFromMatrix(countData = gcount_filt_pln,
                              colData = treatmentinfo,
                              design = ~treatment)
```

#### Visualize gene count data

We're looking to see if the samples of the same pH treatments cluster

##### Log-transform the count data

Log-transform the data using a variance stabilizing transforamtion (vst) for visualization purposes, only. To do this we first need to calculate the size factors of our samples. This is a rough estimate of how many reads each sample contains compared to the others. In order to use VST (the faster log2 transforming process) to log-transform our data, the size factors need to be less than 4. Otherwise, there could be artefacts in our results.
```{r}
SF.gdds_pln <- estimateSizeFactors( gdds_pln ) #estimate size factors to determine if we can use vst  to transform our data. Size factors should be less than for to use vst
print(sizeFactors(SF.gdds_pln)) #View size factors
```

Our size factors are all less than 4, so we can use VST!
```{r}
gvst_pln <- varianceStabilizingTransformation(gdds_pln, blind=TRUE) #apply a variance stabilizing transforamtion to minimize effects of small counts and normalize wrt library size
```

#### Principal component plot of samples
```{r}
gPCAdata_pln <- plotPCA(gvst_pln, intgroup = c("treatment"), returnData=TRUE)
percentVar_pln <- round(100*attr(gPCAdata_pln, "percentVar")) #plot PCA of samples with all data
ggplot(gPCAdata_pln, aes(PC1, PC2, color=treatment)) + 
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar_pln[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar_pln[2],"% variance")) +
  scale_color_manual(values = c(amb="slateblue", l="indianred3", xl="deeppink4")) +
  coord_fixed() +
    theme_bw() + #Set background color
  theme(panel.border = element_blank(), # Set border
                     #panel.grid.major = element_blank(), #Set major gridlines
                     #panel.grid.minor = element_blank(), #Set minor gridlines
                     axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) + #Set the plot background
  theme(legend.position = ("none")) #set title attributes
```

There appears to be clustering by treatment, especially on the PC1 axis.

#### Differential Gene Expression Analysis

##### Run DE analysis

Run differential expression test using a Wald model. 
```{r, message = FALSE}
DEG_pln <- DESeq(gdds_pln) #run differential expression test by group using the Wald model
```

Explore significant p-values for treatment_l_vs_amb, treatment_xl_vs_amb, and treatment_xl_vs_l
```{r, message = FALSE}
DEG_pln.results.l_vs_amb <- results(DEG_pln, contrast=c("treatment","l","amb"))
summary(DEG_pln.results.l_vs_amb) #view results summary
sum(DEG_pln.results.l_vs_amb$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

DEG_pln.results.xl_vs_amb <- results(DEG_pln, contrast=c("treatment","xl","amb"))
summary(DEG_pln.results.xl_vs_amb) #view results summary
sum(DEG_pln.results.xl_vs_amb$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

DEG_pln.results.xl_vs_l <- results(DEG_pln, contrast=c("treatment","xl","l"))
summary(DEG_pln.results.xl_vs_l) #view results summary
sum(DEG_pln.results.xl_vs_l$padj < 0.05, na.rm=TRUE) #How many adjusted p-values were less than 0.05?

#How many adjusted p-values were less than 0.05 and what were they?
LvAMB <- filter(as.data.frame(DEG_pln.results.l_vs_amb), padj<0.05)
XLvAMB <- filter(as.data.frame(DEG_pln.results.xl_vs_amb), padj<0.05)
XLvL <- filter(as.data.frame(DEG_pln.results.xl_vs_l), padj<0.05)

LvAMB$contrast <- "LvAMB"
LvAMB$gene_id <- rownames(LvAMB)

XLvAMB$contrast <- "XLvAMB"
XLvAMB$gene_id <- rownames(XLvAMB)
rownames(XLvAMB) <- seq(from = 13, to = 29, by = 1)
blast_DEGs <- bind_rows(LvAMB, XLvAMB)
```
There are a fews DEGs in the symbionts at the planula stage.

Get fasta for blast
```{r}
fasta <- read.fasta(file = "6-Symbiont-RNAseq-Analysis/a-QC-Align-Assemble/Output/symb_Trinity.fasta", clean_name = FALSE) #load fasta file
fasta$seq.name <- gsub(" ", "", fasta$seq.name)
fasta$seq.name <- as.factor(fasta$seq.name)
```

Get up and down genes for each contrast
```{r}
#Get up and down genes for each contrast
XA.u <- filter(XLvAMB, log2FoldChange>0)
#XA.d <- filter(XLvAMB, log2FoldChange<0) #NO down in XA contrast
LA.u <- filter(LvAMB, log2FoldChange>0)
LA.d <- filter(LvAMB, log2FoldChange<0)

#Extraction
XA.u.seq <- na.omit(fasta[XA.u$gene_id,])
LA.u.seq <- na.omit(fasta[LA.u$gene_id,])
LA.d.seq <- na.omit(fasta[LA.d$gene_id,])
XLLA.u.seq <- na.omit(fasta[intersect(XA.u.seq$seq.name, LA.u.seq$seq.name), ]) #Shared DEGs

#Reformat DFs for saving
rownames(XA.u.seq) <- NULL
XA.u.seq$seq.name <- as.character(XA.u.seq$seq.name)
rownames(LA.u.seq) <- NULL
LA.u.seq$seq.name <- as.character(LA.u.seq$seq.name)
rownames(LA.d.seq) <- NULL
LA.d.seq$seq.name <- as.character(LA.d.seq$seq.name)
rownames(XLLA.u.seq) <- NULL
XLLA.u.seq$seq.name <- as.character(XLLA.u.seq$seq.name)
```

Sav fasta files for BLASTX
```{r}
dat2fasta(XA.u.seq, "6-Symbiont-RNAseq-Analysis/b-Differential-Gene-Expression-Analysis/Output/XAup.fasta")
dat2fasta(LA.u.seq, "6-Symbiont-RNAseq-Analysis/b-Differential-Gene-Expression-Analysis/Output/LAup.fasta")
dat2fasta(LA.d.seq, "6-Symbiont-RNAseq-Analysis/b-Differential-Gene-Expression-Analysis/Output/LAdown.fasta")
dat2fasta(XLLA.u.seq, "6-Symbiont-RNAseq-Analysis/b-Differential-Gene-Expression-Analysis/Output/SHAREDXLLAup.fasta")
```