---
title: "Planula GO Term Enrichment Analysis"
author: "Erin Chille"
date: "2021/02/16"
output: html_document
---

Load necessary libraries
```{r}
library(goseq)
library(tidyverse)
library(ggrepel)
```

Import the data files 
```{r}
#treatment information
treatmentinfo <- read.csv("Sample_Info/RNAseq_data.csv", header = TRUE, sep = ",")
str(treatmentinfo)
head(treatmentinfo)

#gene count matrix
gcount <- as.data.frame(read.csv("2-QC-Align-Assemble/Output/gene_count_matrix.csv", row.names="gene_id"))
gcount <- gcount[ ,treatmentinfo$sample_id] #remove samples not in the treatment information csv (i.e. not in experiment)
gcount$gene_id <- rownames(gcount)
rownames(gcount) <- NULL
dim(gcount)

#DESeq2 and clustering results
DEG.res <- read.csv("4-Differential-Gene-Expression-Analysis/Output/pln_DEGs_all.csv")[,-1]
nrow(DEG.res)

#filter DEGs for log2FoldChange>|1|
DEG.res <- filter(DEG.res, log2FoldChange > 1 | log2FoldChange < (-1))
nrow(DEG.res)

#transcript annotations
annot <- read_tsv( "1-BLAST-GO-KO/Output/200824_Mcap_Blast_GO_KO.tsv", col_names = TRUE) #biological annotation information
Go.ref <- subset(annot, select= c(gene_id, length)) #Select only relevant information

#Filter gcount by available annotations
Go.ref <- merge(gcount, Go.ref, by = "gene_id")
```

Make a dataframe containing the gene_ids and cluster for each cluster.
```{r}
#Select only gene_id and cluster from DEseq2 res
DEGclust <- subset(DEG.res, select=c(gene_id, cluster))
DEGclust <- unique(DEGclust)

clust1 <- filter(DEGclust, cluster=="1")
nrow(clust1) #nrow clust1
clust2 <- filter(DEGclust, cluster=="2")
nrow(clust2) #nrow clust2
```

Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
```{r}
#Make ID and length vectors
Go.ref <- unique(Go.ref)
dim(Go.ref)
IDvector <- Go.ref$gene_id
lengthVector <- Go.ref$length

#Cluster 1
clust1.genes <- as.vector(clust1$gene_id)
clust1.genes=as.integer(Go.ref$gene_id%in%clust1.genes)
names(clust1.genes)=Go.ref$gene_id
length(clust1.genes)
length(names(clust1.genes))
length(unique(names(clust1.genes)))

#Cluster 2
clust2.genes <- as.vector(clust2$gene_id)
clust2.genes=as.integer(Go.ref$gene_id%in%clust2.genes)
names(clust2.genes)=Go.ref$gene_id
length(clust1.genes)

pwf.C1<-nullp(DEgenes=clust1.genes, id=IDvector, bias.data=lengthVector) #weight vector by length of gene
pwf.C2<-nullp(clust2.genes, IDvector, bias.data=lengthVector) #weight vector by length of gene
```

Prepare GO term dataframe
```{r}
GO.annot <- subset(annot, select=c(gene_id, GO_IDs))
GO.annot.na <- filter(GO.annot, GO_IDs!="NA;NA") #Remove NAs
GO.annot.na$GO_IDs <- gsub("NA;", "", GO.annot.na$GO_IDs)  #Remove NAs
GO.annot.na$GO_IDs <- gsub(";NA", "", GO.annot.na$GO_IDs)  #Remove NAs
splitted <- strsplit(as.character(GO.annot.na$GO_IDs), ";") #split into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GO.annot.na$gene_id, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
colnames(GO.terms) <- c("gene_id", "GO.ID")

GO.terms$GO.ID<- as.character(GO.terms$GO.ID)
GO.terms$GO.ID <- replace_na(GO.terms$GO.ID, "unknown")
GO.terms$GO.ID <- as.factor(GO.terms$GO.ID)
GO.terms$gene_id <- as.factor(GO.terms$gene_id)
GO.terms$GO.ID <- gsub(" ", "", GO.terms$GO.ID)
GO.terms <- unique(GO.terms)

dim(GO.terms)
head(GO.terms, 10)

nrow(GO.terms)/length(unique(GO.terms$gene_id)) #avg GO IDs per gene
```

Find enriched GO terms, "selection-unbiased testing for category enrichment amongst significantly expressed genes for RNA-seq data"
```{r, warning=FALSE, message=FALSE}
GOwall.C1 <- goseq(pwf.C1, GOref$gene_id, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GOwall.C2 <- goseq(pwf.C2, GOref$gene_id, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
```

Find only enriched GO terms that are statistically significant at cutoff
```{r}
C1.GO.05<-GOwall.C1$category[GOwall.C1$over_represented_pvalue<.05]
C1.GO.05<-data.frame(C1.GO.05)
colnames(C1.GO.05) <- c("category")
C1.GO.05 <- merge(C1.GO.05, GOwall.C1, by="category")
C1.GO.05 <- C1.GO.05[order(C1.GO.05$ontology, C1.GO.05$over_represented_pvalue, -C1.GO.05$numDEInCat),]
C1.GO.05$term <- as.factor(C1.GO.05$term)
nrow(filter(C1.GO.05, ontology=="BP")) #number sig BP terms
nrow(filter(C1.GO.05, ontology=="MF")) #number sig MF terms
nrow(C1.GO.05)

C2.GO.05<-GOwall.C2$category[GOwall.C2$over_represented_pvalue<.05]
C2.GO.05<-data.frame(C2.GO.05)
colnames(C2.GO.05) <- c("category")
C2.GO.05 <- merge(C2.GO.05, GOwall.C2, by="category")
C2.GO.05 <- C2.GO.05[order(C2.GO.05$ontology, C2.GO.05$over_represented_pvalue, -C2.GO.05$numDEInCat),]
C2.GO.05$term <- as.factor(C2.GO.05$term)
nrow(filter(C2.GO.05, ontology=="BP")) #number sig BP terms
nrow(filter(C2.GO.05, ontology=="MF")) #number sig MF terms
nrow(C2.GO.05)
```

Correct any un-annotated terms/ontologies
```{r}
NAs.ontology <- C1.GO.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- C2.GO.05 %>% subset(is.na(term))
print(NAs.ontology)
```
*There were no un-annotated functions except genes with unknown functions*

Save significant terms
```{r, warning=FALSE}
write.csv(C1.GO.05, file = "5-Planula-GO-Enrichment-Analysis/Output/GO.05.C1.csv", row.names = FALSE)
write.csv(C2.GO.05, file = "5-Planula-GO-Enrichment-Analysis/Output/GO.05.C2.csv", row.names = FALSE)
```

Make dataframe of GO results for clustering and heatmap. Using R script from Mass Lab. Needs columns "experiment" where we will put cluster information, "term", "percentDEInCat", and "gene" with all the DE genes associated with that GO term.
```{r}
C1_fig <- filter(C1.GO.05, ontology=="BP") #separate out BP terms in cluster
C1_fig$experiment <- "down" #label them with "experiment"

C2_fig <- filter(C2.GO.05, ontology=="BP")  #separate out BP terms in cluster
C2_fig$experiment <- "up" #label them with "experiment"

GOclust <- bind_rows(C1_fig, C2_fig) #bind into 1 DF

GOclust$percentDEInCat <- (GOclust$numDEInCat/GOclust$numInCat)*100 #calculate %DEInCat
GOclust <- GOclust[ ,c(8:9,6,1)] #arrange cols

#add gene_IDs. To get gene_IDs we will merge with the GO.terms DF.
GOclust <- rename(GOclust, GO.ID = category) #First have to make the "by" column the same for both
GOclust <- left_join(GOclust, GO.terms, by="GO.ID" ) #join the DFs
GOclust$gene_id <- as.character(GOclust$gene_id) #make gene ID a character so we can collasp our many near-identical columns

GOclust <- GOclust %>% #collapse and have gene IDs for a particular term in a single row as a comma-sep list. 
  group_by(experiment, percentDEInCat, term) %>%
  summarise(genes = toString(gene_id)) %>% #rename collapsed gene_ID column "gene"
  ungroup()
write.csv(GOclust, file = "5-Planula-GO-Enrichment-Analysis/Output/GO.05.clust.csv", row.names = FALSE)
```
Make dot plot
```{r}
GOdot <- bind_rows(C1_fig, C2_fig) #bind into 1 DF
GOdot$percentDEInCat <- (GOdot$numDEInCat/GOdot$numInCat)*100 #calculate %DEInCat
GOdot$over_represented_pvalue <- -log10(GOdot$over_represented_pvalue) #-log10 p-value for plotting
GOdot <- rename(GOdot, direction=experiment) #rename experiment col to direction of low compared to xlow and amb

# plot: dot plot
GOdot.plot <- GOdot %>% ggplot(aes(x = reorder(term, percentDEInCat), y = percentDEInCat, 
  color = `direction`, size = over_represented_pvalue)) + 
  coord_flip() +
  geom_point() +
  scale_color_manual(values=c(up = "red", down = "blue")) +
  theme_bw() + 
  ylab("percentDEInCat") + 
  xlab("term") +
  theme(axis.line = element_line(color = 'black'), #Set the axes color
        panel.border = element_blank(), #Set the border
        axis.text.x=element_text(size = 9), #set y-axis label size
        axis.title.x=element_text(size =10), #set y-axis title size
        axis.text.y=element_text(size = 9), #set x-axis label size
        legend.text=element_text(size=9), #set legend text size
        legend.title=element_text(size=10), #set legend title size
        axis.title.y=element_text(size = 10)) #set x-axis title size
  labs(size = "-log10(p-value)"); GOdot.plot # re-label size labels to reflect p-value tranformation

ggsave(GOdot.plot, filename = "5-Planula-GO-Enrichment-Analysis/Output/GO.05.dotplot.png",
  width = 8,
  height = 7,
  units = c("in"),
  dpi = 300)
```



